pkg/disk/groupcontrollerserver.go

```diff
@@ -0,0 +1,287 @@
+	var caps []*csi.GroupControllerServiceCapability
+	for _, cap := range groupControllerCaps {
+		c := &csi.GroupControllerServiceCapability{
+			Type: &csi.GroupControllerServiceCapability_Rpc{
+				Rpc: &csi.GroupControllerServiceCapability_RPC{
+					Type: cap,
+				},
+			},
+		}
+		caps = append(caps, c)
+	}
+	return &csi.GroupControllerGetCapabilitiesResponse{Capabilities: caps}, nil
+}
+
+func (cs *groupControllerServer) CreateVolumeGroupSnapshot(ctx context.Context, req *csi.CreateVolumeGroupSnapshotRequest) (*csi.CreateVolumeGroupSnapshotResponse, error) {
+	// request limit
+	cur := time.Now().Unix()
+	if initTime, ok := GroupSnapshotRequestMap[req.Name]; ok {
+		if cur-initTime < GroupSnapshotRequestInterval {
+			return nil, status.Errorf(codes.Aborted, "volume group snapshot create request limit %s", req.Name)
+		}
+	}
+	GroupSnapshotRequestMap[req.Name] = cur
+
+	// used for snapshot events
+	groupSnapshotName := req.Parameters[common.VolumeGroupSnapshotNameKey]
+	groupSnapshotNamespace := req.Parameters[common.VolumeGroupSnapshotNamespaceKey]
+
+	ref := &v1.ObjectReference{
+		Kind:      "VolumeGroupSnapshot",
+		Name:      groupSnapshotName,
+		UID:       "",
+		Namespace: groupSnapshotNamespace,
+	}
+
+	klog.Infof("CreateVolumeGroupSnapshot:: Starting to create volumegroupsnapshot: %+v", req)
+
+	sourceVolumeIds := req.GetSourceVolumeIds()
+	for index, id := range sourceVolumeIds {
+		sourceVolumeIds[index] = strings.TrimSpace(id)
+	}
+	// Need to check for already existing snapshot name
+	GlobalConfigVar.EcsClient = updateEcsClient(GlobalConfigVar.EcsClient)
+	// check if groupvolumesnapshot has already exists
+	existsGroupSnapshot, err := findSnapshotGroup(req.GetName(), "")
+
+	// case 1: groupSnapshot with same name existes, but is failed or inprogress
+	if err != nil {
+		klog.Errorf("CreateVolumeGroupSnapshot::  Snapshot already created: name[%s], and get error: %v", req.GetName(), err)
+		e := status.Errorf(codes.Internal, "CreateVolumeGroupSnapshot: volumeGroupSnapshot with the same name: %s exists but with error: %s", req.GetName(), err.Error())
+		utils.CreateEvent(cs.recorder, ref, v1.EventTypeWarning, snapshotCreateError, e.Error())
+		return nil, e
+	}
+
+	// case 2: groupSnapshot with same name existes, and is accomplished
+	if existsGroupSnapshot != nil {
+		// check if snapshots in group match sourceVolumeIds for an accomplish volume group
+		match := ifExistsGroupSnapshotMatchSourceVolume(existsGroupSnapshot, sourceVolumeIds)
+		if !match {
+			klog.Errorf("CreateVolumeGroupSnapshot:: GroupSnapshot already exist with same name: name[%s], but with different SourceVolumeIDs[%v]", req.Name, sourceVolumeIds)
+			err := status.Errorf(codes.AlreadyExists, "groupSnapshot with the same name: %s but with different SourceVolumeIds already exist", req.GetName())
+			utils.CreateEvent(cs.recorder, ref, v1.EventTypeWarning, snapshotAlreadyExist, err.Error())
```

Comment:
it will create event on snapshot content, i think it is a same issue for single snapshot in controllerserver, i will submit another PR after this one is mergerd.